File: sorting-algos/divide-and-conquer/merge-sort.py
def merge_sort(arr):
    
    helper(arr, 0, len(arr) - 1)

    return arr

def helper(arr, start, end):

    if start == end:
        return arr

    
    mid = (start + end) // 2

    print("START: "+ str(start))
    print("END: "+ str(end))
    print("MID: "+ str(mid))

    helper(arr, start, mid)
    helper(arr, mid + 1, end)

    i = start
    j = mid + 1
    aux = []
    
    while i <= mid and j <= end:
        if arr[i] <= arr[j]:
            aux.append(arr[i])
            i += 1
        else:
            aux.append(arr[j])
            j += 1

    while i <= mid:
        aux.append(arr[i])
        i += 1

    while j <= end:
        aux.append(arr[j])
        j += 1

    arr[start:end] = aux

def main():
    # big_array = [587, 476, 54, 767, 982, 895, 353, 55, 508, 230, 505, 990, 189, 555, 460, 665, 432, 24, 321, 399, 643, 167, 828, 371, 293, 251, 466, 67, 533, 908, 477, 509, 39, 407, 935, 917, 356, 196, 717, 443, 180, 924, 898, 88, 385, 363, 490, 386, 276, 947, 963, 525, 403, 876, 727, 277, 80, 190, 117, 734, 861, 411, 100, 565, 944, 500, 484, 14, 89, 806, 965, 96, 795, 364, 663, 816, 416, 995, 664, 512, 608, 295, 943, 933, 784, 286, 462, 641, 929, 480, 926, 639, 632, 548, 280, 818, 650, 911, 571, 226]
    small_array = [7, 5, 4, 6, 0, 8, 3, 1, 2, 9]
    # print(insertion_sort(big_array))
    print(merge_sort(small_array))
    # return [4, 4, 4, 4, 5, 5, 5, 5, 6, 7, 6, 6, 6, 0, 8, 3, 1, 2, 9]


main()

File: sorting-algos/divide-and-conquer/merge-sort.swift
func merge_sort(arr: [Int]) -> [Int] {
    
    guard arr.count > 1 else { return arr }

    let middleIndex = arr.count / 2

    let leftArray = merge_sort(arr: Array(arr[0..<middleIndex]))
    let rightArray = merge_sort(arr: Array(arr[middleIndex..<arr.count]))

    return merge(leftArray: leftArray, rightArray: rightArray)
}

func merge(leftArray: [Int], rightArray: [Int]) -> [Int] {
    var leftIndex = 0
    var rightIndex = 0

    var orderedArray: [Int] = []

    while leftIndex < leftArray.count && rightIndex < rightArray.count {
        let leftElement = leftArray[leftIndex]
        let rightElement = rightArray[rightIndex]

        if leftElement == rightElement {
            orderedArray.append(contentsOf: [leftElement, rightElement])
            leftIndex += 1
            rightIndex += 1
        } else if leftElement <= rightElement {
            orderedArray.append(leftElement)
            leftIndex += 1
        } else {
            orderedArray.append(rightElement)
            rightIndex += 1
        }
    }

    while leftIndex < leftArray.count {
        orderedArray.append(leftArray[leftIndex])
        leftIndex += 1
    }

    while rightIndex < rightArray.count {
        orderedArray.append(rightArray[rightIndex])
        rightIndex += 1
    }

    return orderedArray
}


let smallArray = [7, 5, 4, 6, 0, 8, 3, 1, 2, 9]
// merge_sort(arr: smallArray)
print("MERGE_SORT(ARR: SMALLARRAY): \(String(describing: merge_sort(arr: smallArray)))")